### PR TITLE
Adds a fmf node validation layer to core classes (Test/Plan/Story)

### PR DESCRIPTION
**Pull request by happz**
Before letting core classes consume data loaded from fmf files, a JSON
schema-based validation is applied to discover possible issues.

Validation is implemented as a mixin class, therefore it is not limited
to these classes and may be re-used by other parts of code.

### CODE REVIEW

**Comment by happz** [comment]
Ready for review.

Contains plenty of apparently unrelated changes in fmf files - these are caused by tmt suddenly using schemas to validate plans and tests it loads, including the plans and tests "hidden" under some tests. These custom plans are invisible to tmt scanning the top-level fmf tree of the repository, but they are still loaded by tmt under the control of some tests. And some of these tests are sensitive to extra info printed out by tmt.

I plan to add a patch to run `test_schemas.py` for all of these "hidden" trees, but the problem is, my changes in this patch are needed first, which introduces a chicken and egg problem: some of the changes of this future "protect even custom fmf trees" patch is already in this PR because I need them to make tests happy so I could introduce changes that would make this future patch possible and easy to develop.

**Comment by happz** [comment]
@lukaszachy please, consider this PR for 1.17. It's a blocker for #1178, which is then blocking mypy from following imports.

**Comment by lukaszachy** [comment]
Hm, I am surprised that tmt doesn't require 'how' but we even have it in the examples - https://tmt.readthedocs.io/en/stable/spec/plans.html#the-simplest-usage

Is it possible to do the validation without 'how' key? This could break user plans (aka 2.0)

**Comment by happz** [comment]
> Hm, I am surprised that tmt doesn't require 'how' but we even have it in the examples - https://tmt.readthedocs.io/en/stable/spec/plans.html#the-simplest-usage
> 
> Is it possible to do the validation without 'how' key? This could break user plans (aka 2.0)

I think it may be possible in some cases, but in general, it may be hard or even impossible. Imagine two plugins with some keys/options supported by both of them (e.g. `arch`), but maybe with different types allowed. Plugin Foo allows `arch` to be any string, plugin Bar limits `arch` to be an enum with `aarch64` and `x86_64` values only. When tmt sees `arch: aarch64` in a `provision` step, it seems really hard to decide which plugin schema to use for validation.

Is the required `how` a serious blocker? IIUIC, `tmt` injects a default `how` when none is found, fixing fmf files should make things right. tmt specification not requiring `how` is indeed a complication...

**Comment by lukaszachy** [comment]
> Is the required `how` a serious blocker? IIUIC, `tmt` injects a default `how` when none is found, fixing fmf files should make things right. tmt specification not requiring `how` is indeed a complication...

@psss ? I'd say better to be explicit but tmt has one or two 'save user typing two words and make it much more complicated to the tmt devel' type of features.

File: plans/install/docs.fmf [hunk 1]
```diff
@@ -6,6 +6,7 @@ prepare:
     how: install
     package: python3-pip
 execute:
+    how: tmt
```
Comment (thrix) [inline plans/install/docs.fmf:9]: @happz no more default `how` ?

Comment (thrix) [inline plans/install/docs.fmf:9]: @happz no more default `how` ?

Comment (thrix) [reply plans/install/docs.fmf:9]: I thought this is auto-added by tmt, thus not needed?

File: tests/unit/test_schemas.py [hunk 2]
```diff
@@ -6,70 +6,73 @@
...
-SCHEMADIR = os.path.join(PATH, "../../tmt/schemas")
 ROOTDIR = os.path.join(PATH, "../..")
 
+TREE = tmt.Tree(ROOTDIR)
 
-@pytest.fixture
-def tests_schema():
-    return tmt.utils.load_schema('test.yaml')
 
+# This is what `Tree.{tests,plans,stories}`` do internally, but after getting
```
Comment (thrix) [inline tests/unit/test_schemas.py:14]: Hmm, I am not sure, I thought default how's added are expected from tmt

@psss @lukaszachy can you pls chime in?

File: tmt/schemas/provision/hardware.yml [hunk 3]
```diff
@@ -0,0 +1,220 @@
+---
```
Comment (thrix) [inline tmt/schemas/provision/hardware.yml:1]: @happz seems not rebased? this file should not be here

**Review by thrix — CHANGES_REQUESTED** [review]
@happz pls rebase, seems this is based on some older change, some files, like `tests.yaml` and `hardware.yml` were renamed afairr

File: plans/install/docs.fmf (hunk 1 above)
Comment (happz) [reply plans/install/docs.fmf:9]: That's correct, Added on today's hacking meeting notes.

File: tmt/schemas/provision/hardware.yml (hunk 3 above)
Comment (happz) [reply tmt/schemas/provision/hardware.yml]: Should be fixed, the history was pretty weird.

**Comment by happz** [comment]
@thrix please, https://github.com/teemtee/tmt/pull/1421 before this one - all those schema changes are worth their own patch.

File: tests/unit/test_schemas.py (hunk 2 above)
Comment (psss) [reply tests/unit/test_schemas.py:14]: Those will be handled by the "prenormalization".

File: tmt/utils.py [hunk 4]
```diff
@@ -2656,6 +2671,143 @@ def load_schema_store() -> SchemaStore:
     return store
 
 
+def _prenormalize_fmf_node(node: fmf.Tree, schema_name: str) -> fmf.Tree:
+    """
+    Apply the minimal possible normalization steps to nodes before validating them with schemas.
+
+    Tmt allows some fields to have default values, and at least ``how`` field is necessary for
```
Comment (psss) [inline tmt/utils.py:2678]:
```suggestion
    tmt allows some fields to have default values, and at least ``how`` field is necessary for
```

File: tmt/utils.py [hunk 5]
```diff
@@ -2656,6 +2671,143 @@ def load_schema_store() -> SchemaStore:
...
+
+        step['how'] = cast(tmt.steps.Step, step_class).DEFAULT_HOW
+
+    def _process_step_collection(step_name: str, step_collection: Any) -> None:
+        """
+        Process a collection of step configurations.
+        """
+
+        # Ignore anything that is not a step.
+        if step_name not in ('discover', 'execute', 'finish', 'prepare', 'provision', 'report'):
```
Comment (psss) [inline tmt/utils.py:2760]:
```suggestion
        if step_name not in tmt.steps.STEPS:
```

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for implementing the validation! Looks good, added just a few minor suggestions.

File: tmt/base.py [hunk 6]
```diff
@@ -107,6 +107,57 @@ def validate(self) -> Tuple[bool, str]:
         return (True, '')
 
 
+class ValidateFmfMixin:
+    """
+    Mixin adding validation of a fmf node.
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/base.py]: Fixed.

File: tmt/base.py [hunk 7]
```diff
@@ -107,6 +107,57 @@ def validate(self) -> Tuple[bool, str]:
...
+            return
+
+        # Cannot use backslash inside f-string, but I want to join errors with a new-line
+        # character, hence this dummy helper variable.
+        NL = '\n'
+
+        message = f"""Failed to validate fmf node {node.name}
+
+{NL.join('* ' + error_message for _, error_message in errors)}
+"""
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/base.py]: Sure, fixed.

File: tmt/utils.py (hunk 4 above)
Comment (happz) [reply tmt/utils.py]: Fixed.

File: tmt/utils.py (hunk 5 above)
Comment (happz) [reply tmt/utils.py]: Nice, fixed.

**Comment by happz** [comment]
Rebased & addressed reported issues.

**Review by psss — APPROVED** [review]
Thanks for all the fixes! Looks good.

**Review by thrix — APPROVED** [review]

**Pull request merged**